plugins/pulp_docker/plugins/importers/upload.py

```diff
@@ -169,6 +179,47 @@ def get_models(self, metadata, mask_id=''):
         return images
 
 
+class ImportManifestList(PluginStep):
+    """
+    Add an uploaded ManifestList to a repository.
+    """
+
+    def process_main(self):
+        """
+        Validate the uploaded manifest list json, then import content unit into repository.
+        """
+        with open(self.parent.file_path, 'r') as uploaded_file:
+            manifest_list = uploaded_file.read()
+        models.ManifestList.validate_json(manifest_list)
+        digest = models.Manifest.calculate_digest(manifest_list)
+        manifest_list_instance = models.ManifestList.from_json(manifest_list, digest)
+        transfer_repo = self.get_repo()
+
+        #  Ensure that all referenced manifests are already in repository.
+        manifest_digests = set()
+        for manifest in manifest_list_instance.manifests:
+            manifest_digests.add(manifest)
+        qs = models.Manifest.objects.filter(
+            digest__in=sorted(manifest_digests)).only('id', 'digest')
+        referenced_manifests = {manifest['digest']: manifest['id'] for manifest in qs}
+        unit_qs = pulp_models.RepositoryContentUnit.objects.filter(
+            repo_id=transfer_repo.id,
+            unit_id__in=referenced_manifests.values()).values_list('unit_id')
+        unit_ids_in_repo = list(unit_qs)
+        for digest in manifest_digests:
+            unit_id = referenced_manifests.get(digest)
+            if not unit_id or unit_id not in unit_ids_in_repo:
+                raise PulpCodedValidationException(error_code=error_codes.DKR1013,
+                                                   digest=digest)
+
+        try:
+            manifest_list_instance.save_and_import_content(self.parent.file_path, digest)
```

Comment:
the storage path is not set here, the content will be saved to None location

Reply:
A pre_save signal for FileContentUnits calls `set_storage_path`, which correctly sets the storage path for the uploaded Manifest List. (proven by the fact that file exists at the _storage_path in the db). This makes sense for Manifest Lists, which are a single file, so we can determine its path within the unit storage by its filename.

Reply:
it took me some time, but for the sake of prooving that we do need to set the storage path in this case:
```
$ curl -L -H "Accept: application/vnd.docker.distribution.manifest.list.v2+json" -X GET -k 'http://localhost:1234/v2/synctest1/manifests/sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0'
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<html>
 <head>
  <title>Index of /pulp/docker/v2/synctest1/manifests/list/sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0</title>
 </head>
 <body>
<h1>Index of /pulp/docker/v2/synctest1/manifests/list/sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0</h1>
  <table>
   <tr><th valign="top"><img src="/icons/blank.gif" alt="[ICO]"></th><th><a href="?C=N;O=D">Name</a></th><th><a href="?C=M;O=A">Last modified</a></th><th><a href="?C=S;O=A">Size</a></th><th><a href="?C=D;O=A">Description</a></th></tr>
   <tr><th colspan="5"><hr></th></tr>
<tr><td valign="top"><img src="/icons/back.gif" alt="[PARENTDIR]"></td><td><a href="/pulp/docker/v2/synctest1/manifests/list/">Parent Directory</a>       </td><td>&nbsp;</td><td align="right">  - </td><td>&nbsp;</td></tr>
<tr><td valign="top"><img src="/icons/unknown.gif" alt="[   ]"></td><td><a href="./sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0">sha256:bbc3a03235220..&gt;</a></td><td align="right">2017-11-23 12:51  </td><td align="right">2.6K</td><td>&nbsp;</td></tr>
   <tr><th colspan="5"><hr></th></tr>
</table>
</body></html>
[ipanova@ina ~]$ sudo su
[root@ina ipanova]# ll /var/lib/pulp/published/docker/v2/master/synctest1/1511438278.63/manifests/list/
total 8
lrwxrwxrwx. 1 apache apache 114 Nov 23 12:57 myname -> /var/lib/pulp/content/units/docker_manifest_list/cf/dbc0f5630ac3ab57e6e255e022339df7afe6a5dfc199ad1eb4665d0f9df7e3
lrwxrwxrwx. 1 apache apache 114 Nov 23 12:57 sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0 -> /var/lib/pulp/content/units/docker_manifest_list/cf/dbc0f5630ac3ab57e6e255e022339df7afe6a5dfc199ad1eb4665d0f9df7e3
[root@ina ipanova]# exit
[ipanova@ina ~]$ docker pull localhost:1234/synctest1:myname
Error response from daemon: invalid character '<' looking for beginning of value
[ipanova@ina ~]$ sudo su
[root@ina ipanova]# less /var/log/httpd/access_log| tail -5
::1 - - [23/Nov/2017:12:59:17 +0100] "GET /v2/synctest1/manifests/myname HTTP/1.1" 302 367 "-" "docker/1.12.6 go/go1.6.4 git-commit/78d1802 kernel/4.11.12-100.fc24.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/1.12.6 \\(linux\\))"
::1 - - [23/Nov/2017:13:01:37 +0100] "GET /v2/synctest1/manifests/sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0 HTTP/1.1" 302 497 "-" "curl/7.47.1"
::1 - - [23/Nov/2017:13:02:13 +0100] "\x16\x03\x01" 400 226 "-" "-"
::1 - - [23/Nov/2017:13:02:13 +0100] "GET /v2/ HTTP/1.1" 200 2 "-" "docker/1.12.6 go/go1.6.4 git-commit/78d1802 kernel/4.11.12-100.fc24.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/1.12.6 \\(linux\\))"
::1 - - [23/Nov/2017:13:02:13 +0100] "GET /v2/synctest1/manifests/myname HTTP/1.1" 302 367 "-" "docker/1.12.6 go/go1.6.4 git-commit/78d1802 kernel/4.11.12-100.fc24.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/1.12.6 \\(linux\\))"
[root@ina ipanova]# exit
[ipanova@ina ~]$ mongo
MongoDB shell version: 3.2.12
connecting to: test
Server has startup warnings: 
2017-10-23T11:05:22.157+0200 I CONTROL  [initandlisten] 
2017-10-23T11:05:22.157+0200 I CONTROL  [initandlisten] ** WARNING: soft rlimits too low. rlimits set to 31757 processes, 64000 files. Number of processes should be at least 32000 : 0.5 times number of files.
> use pulp_database
switched to db pulp_database
> db.units_docker_manifest_list.find({"digest":"sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0"})[0]
{
	"_id" : "0f74d2f0-b6c7-40a3-8070-27d8b18ffab4",
	"pulp_user_metadata" : {
		
	},
	"_last_updated" : 1511437865,
	"_storage_path" : "/var/lib/pulp/content/units/docker_manifest_list/cf/dbc0f5630ac3ab57e6e255e022339df7afe6a5dfc199ad1eb4665d0f9df7e3",
	"downloaded" : true,
	"digest" : "sha256:bbc3a03235220b170ba48a157dd097dd1379299370e1ed99ce976df0355d24f0",
	"schema_version" : 2,
	"manifests" : [
		"sha256:91ef6c1c52b166be02645b8efee30d1ee65362024f7da41c404681561734c465",
		"sha256:0be57da4f47fe310100c465f7be198b54c957c31a58e9bc390bbfda29003ec36",
		"sha256:45e6a3e59a8a1af3fb93d68ee4b415fe8f6b0b1fcf75d13926f1a529f556e084",
		"sha256:2813337ff358245e1c8a54c58050e978ae753e67c20093ff4ffd67c205d6b732",
		"sha256:f814888b4bb6149bd39ba8375a1932fb15071b4dbffc7f76c7b602b06abbb820",
		"sha256:c87a380a858bc39939822340be96ff4fd8d6ecb8b3489fa08ac88267932e87bf",
		"sha256:5414799e8dc5f463617e21950f94f93fca0d8a2d939dae62188b0aa0915c4a13",
		"sha256:bf39f0f063f21da82330e0e7eec95a2eaa3ed80f42f71bb9e08a274131cbeb9c"
	],
	"amd64_digest" : "sha256:91ef6c1c52b166be02645b8efee30d1ee65362024f7da41c404681561734c465",
	"amd64_schema_version" : 2,
	"_ns" : "units_docker_manifest_list",
	"_content_type_id" : "docker_manifest_list"
}
> 

```
_storage_path in DB points to the dir and not the file
during publish symlink created points to a directly that as consequence prevents to make successful docker pull as well as curl request.

Reply:
to fix this you need:
```
[ipanova@ina pulp_docker]$ git diff
diff --git a/plugins/pulp_docker/plugins/importers/upload.py b/plugins/pulp_docker/plugins/importers/upload.py
index b781271..675d1f8 100644
--- a/plugins/pulp_docker/plugins/importers/upload.py
+++ b/plugins/pulp_docker/plugins/importers/upload.py
@@ -177,7 +177,8 @@ class ProcessMetadata(PluginStep):
 
         return images
 
-
+import logging
+_logger = logging.getLogger(__name__)
 class AddManifestList(PluginStep):
     """
     Add an uploaded ManifestList to a repository.
@@ -213,8 +214,9 @@ class AddManifestList(PluginStep):
             raise PulpCodedValidationException(error_code=error_codes.DKR1013,
                                                digests=missing_manifest_digests)
 
+        manifest_list_instance.set_storage_path(digest)
         try:
-            manifest_list_instance.save_and_import_content(self.parent.file_path, digest)
+            manifest_list_instance.save_and_import_content(self.parent.file_path)
         except NotUniqueError:
             manifest_list_instance = models.ManifestList.objects.get(digest=digest)
         repository.associate_single_unit(transfer_repo.repo_obj, manifest_list_instance)
```